packages/compose/src/hooks/use-drop-zone/index.js

```diff
@@ -138,12 +138,38 @@ export default function useDropZone( {
 				}
 			}
 
+			/** @type {number | undefined} */
+			let lastDragX;
+			/** @type {number | undefined} */
+			let lastDragY;
+			/** @type { number | undefined } */
+			let lastTime; // Note that FF will reduce precision to 100ms when privacy.resistFingerprinting is enabled
 			function onDragOver( /** @type {DragEvent} */ event ) {
 				// Only call onDragOver for the innermost hovered drop zones.
 				if ( ! event.defaultPrevented && onDragOverRef.current ) {
-					onDragOverRef.current( event );
+					const time = Date.now();
+					if (
+						lastTime === undefined ||
+						lastDragX === undefined ||
+						lastDragY === undefined
+					) {
+						//first frame
+						lastTime = time;
+						lastDragX = event.clientX;
+						lastDragY = event.clientY;
+					}
+					if ( time - lastTime > 33 ) {
+						const distance =
+							Math.abs( event.clientY - lastDragY ) +
+							Math.abs( event.clientX - lastDragX );
+						if ( distance < 10 ) {
+							onDragOverRef.current( event );
+						}
+						lastDragX = event.clientX;
+						lastDragY = event.clientY;
+						lastTime = time;
+					}
```

Comment:
What would be the difference with lodash's throttle? If we keep this type of throttling, any change we can abstract this out of this hook?

Reply:
> What would be the difference with lodash's throttle?

Great question! So using some of the nice visualizations from: https://css-tricks.com/debouncing-throttling-explained-examples/

#### Debounce

<img width="691" alt="debounce" src="https://user-images.githubusercontent.com/1270189/127025596-00be8c36-19ff-4371-aedd-e52f5e4f8b2a.png">

Debounce turns a number of raw events into a single (usually trailing) event call for some interval. This is really great for things like search inputs/API calls. Say one API call to search for "cats" instead of four API requests for "c", "ca", "cat", "cats" (which might resolve in incorrect order).

#### Throttle 

<img width="765" alt="throttle" src="https://user-images.githubusercontent.com/1270189/127025566-e47cf005-411f-4e3d-8ae6-3dad46c58216.png">

With throttle we don’t allow to our function to execute more than once every X milliseconds. This usually works well with things involving animation, since we'd like to see all the inbetween work at a consistent rate. For example, imagine if a scroll handler skipped all the way to the top or bottom using debounce without animating the inbetween frames, it'd be jarring.

#### Heuristic - Is the user dragging slowly?

In this case, I'm using a pretty basic heuristic of is the user dragging slowly? If they are, we can guess that they intend to drop an item soon and so we should show drag inserters/hints (with all the fancy expensive animations that come with that). If not we can skip work.

What's tricky is **tuning** the **sensitivity**. If it's too sensitive we'll show the drag inserters too often (and slow down dragging), if it's not sensitive enough users will have trouble dropping items into dropzones.

A great way of visualizing this is with an old jQuery plugin hover intent that uses mouse velocity to trigger if a hover event should fire or not https://briancherne.github.io/jquery-hoverIntent/:

See how moving the mouse back and forth in the first row can queue up a lot of events, while the second row is closer to what a user would expect for interactive behavior?

https://user-images.githubusercontent.com/1270189/127026591-32ca04c6-027f-487d-84dd-84555960bdaa.mp4

> If we keep this type of throttling, any change we can abstract this out of this hook?

We can certainly pull out a mouse velocity/user intent piece utility with sensitivity/timing options, but I'd maybe hold off until we have a second example. It looks like it's somewhat rare for us to need to tune raw event handling like this. 

If you know that we already have a user intent/mouse velocity library in place already, I'm happy to update the implementation too.

Reply:
Sure, I'll see if I we can tidy things a bit more without affecting cost too much

Reply:
@ellatrix were you thinking of something like https://github.com/WordPress/gutenberg/pull/33435/commits/2bfbe2368629b1524416fef2066c29fc474b6add or https://github.com/WordPress/gutenberg/pull/33435/commits/b280bb0a3a11b6df84b40dd8b30278333497636d ?

Reply:
This is a great comment https://github.com/WordPress/gutenberg/pull/33435/files#r676775855 that I think deserves to be an inline documentation or an architecture doc.

Is this behavior similar to the drag and scroll behavior we have where we also compute mouse velocity? Could this be considered a second use-case for the same utility.

Reply:
> I think deserves to be an inline documentation or an architecture doc.

Sure, happy to add one. Maybe a subsection for common strategies in https://github.com/WordPress/gutenberg/blob/trunk/docs/explanations/architecture/performance.md ?

> Is this behavior similar to the drag and scroll behavior we have where we also compute mouse velocity? Could this be considered a second use-case for the same utility.

If we're computing mouse velocity elsewhere, we can potentially reuse it for this handler. @youknowriad do you have a link to the file for where this happens? I can try incorporating it here or in a follow up.

Reply:
> Sure, happy to add one. Maybe a subsection for common strategies in https://github.com/WordPress/gutenberg/blob/trunk/docs/explanations/architecture/performance.md ?

Sounds good to me.

> If we're computing mouse velocity elsewhere, we can potentially reuse it for this handler. @youknowriad do you have a link to the file for where this happens? I can try incorporating it here or in a follow up.

I'm not very familiar with this hook but I think this is where we're kind of doing similar things:

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-draggable/use-scroll-when-dragging.js

The idea of that hook is to scroll the container as we get to the edge with the mouse but there's something about velocity as well, if you're moving fast, it scrolls faster.

Reply:
I still think this is worth separating for clarity, ideally just a wrapper function similar to throttle and debounce that the implementor of the drop zone can add around the callbacks. useDropZone is otherwise a simple utility around the native drag and drop API to make it easier to work with, it shouldn't have any opinions about the rest imho.